src/main/java/org/folio/circulation/support/MultipleRecordFetcher.java

```diff
@@ -28,20 +33,58 @@ public MultipleRecordFetcher(
     this.recordMapper = recordMapper;
   }
 
-  public CompletableFuture<Result<MultipleRecords<T>>> findByIds(Collection<String> idList) {
-    if(idList.isEmpty()) {
+  public CompletableFuture<Result<MultipleRecords<T>>> findByIds(
+      Collection<String> ids) {
+    return findByIndexName(ids, "id");
+  }
+
+  public CompletableFuture<Result<MultipleRecords<T>>> findByIndexName(
+      Collection<String> ids, String indexName) {
+    if (ids.isEmpty()) {
       return completedFuture(of(MultipleRecords::empty));
     }
-
-    return findByIndexName(idList, "id");
+    return findByBatchQueries(buildBatchQueriesByIndexName(ids, indexName));
   }
 
-  public CompletableFuture<Result<MultipleRecords<T>>> findByIndexName(Collection<String> idList, String indexName) {
-    if(idList.isEmpty()) {
+  public CompletableFuture<Result<MultipleRecords<T>>> findByIndexNameAndQuery(
+      Collection<String> ids, String indexName, Result<CqlQuery> andQuery) {
+    if (ids.isEmpty()) {
       return completedFuture(of(MultipleRecords::empty));
     }
+    return findByBatchQueriesAndQuery(buildBatchQueriesByIndexName(ids, indexName), andQuery);
+  }
 
-    return findByQuery(exactMatchAny(indexName, idList));
+  private CompletableFuture<Result<MultipleRecords<T>>> findByBatchQueriesAndQuery(
+      List<Result<CqlQuery>> queries, Result<CqlQuery> andQuery) {
+    return findByBatchQueries(queries.stream().map(query ->
+      andQuery.combine(query, CqlQuery::and))
+      .collect(Collectors.toList()));
+  }
+  
+  private List<Result<CqlQuery>> buildBatchQueriesByIndexName(
+      Collection<String> ids, String indexName) {
+    List<Result<CqlQuery>> queries = new ArrayList<>();
+    List<String> idsList = new ArrayList<>(ids);
+    while (!idsList.isEmpty()) {
+      int currentSize = idsList.size() <= MAX_BATCH_SIZE ? idsList.size() : MAX_BATCH_SIZE;
+      List<String> currentIds = idsList.subList(0, currentSize);
+      queries.add(exactMatchAny(indexName, currentIds));
+      idsList = idsList.subList(currentSize, idsList.size());
+    }
+    return queries;
+  }
+
+  private CompletableFuture<Result<MultipleRecords<T>>> findByBatchQueries(
+      List<Result<CqlQuery>> queries) {
+    // NOTE: query limit is max value to ensure all records are returned
+    List<CompletableFuture<MultipleRecords<T>>> results = queries.stream()
+        .map(query -> findByQuery(query, Integer.MAX_VALUE)
+        .thenApply(result -> result.value()))
```

Comment:
What happens if one or more of these requests fail?

Reply:
Not sure, you tell me. What will happen if and when the single request for default limit failed before.

Reply:
Apologies I think I may have asked the question poorly, without enough context.

I try to ask question in order to encourage thinking about situations. I'm aware that sometimes I do that in a way that is less than helpful.

I don't know specifically what the single request situation did. I imagine it propagated the failure back to the client.

I asked the question because this combining of multiple responses together assumes that all of them succeed. The line `.thenApply(result -> result.value()))` gets the value of the result, which will be null when a failure has occurred.

This will likely cause a null pointer exception during the line `.map(records -> (MultipleRecords<T>) records.join())` that might result in no response being sent back to the client of mod-circulation at all.

Reply:
I see. I will update accordingly.

Reply:
Hmm, looking at the code, my interpretation of where the null pointer exception could occur is likely wrong

Reply:
Will there be an unhandled exception? `findByQuery` calls MultipleRecords.from, which performs `ResponseInterpreter flatMapOn 200`. If we chain those `CompletableFuture` together will each have values in the results?